ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestCase.java

```diff
@@ -295,4 +314,208 @@ private static String standardizeTopology(final String topology) {
+      final Object value;
+      if (expected.get(i).getTestRecord().topic.getSerdeSupplier()
+          instanceof ValueSpecJsonSerdeSupplier) {
+        value = new String((byte[]) expectedProducerRecord.value(), Charset.forName("UTF-8"));
+      } else {
+        final Deserializer deserializer = expected.get(i).getTestRecord().topic
+            .getDeserializer(schemaRegistryClient);
+        value = deserializer.deserialize("", (byte[]) expectedProducerRecord.value());
+      }
+
+      if (
+          !actualProducerRecord.timestamp().equals(expectedProducerRecord.timestamp())
+          || !actualProducerRecord.key().equals(expectedProducerRecord.key())
+          || !actualProducerRecord.value().equals(value)
+          ) {
+        final AssertionError error = new AssertionError(
+            "Expected <" + expectedProducerRecord.key() + ", "
+                + value + "> with timestamp="
+                + expectedProducerRecord.timestamp()
+                + " but was <" + actualProducerRecord.key() + ", "
+                + actualProducerRecord.value() + "> with timestamp="
+                + actualProducerRecord.timestamp());
+        error.printStackTrace();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Map<String, List<FakeKafkaRecord>> getExpectedRecordsMap(
+      final SchemaRegistryClient schemaRegistryClient) {
+    final Map<String, List<FakeKafkaRecord>> outputRecordsFromTest = new HashMap<>();
+    outputRecords.forEach(
+        record ->  {
+          if (!outputRecordsFromTest.containsKey(record.topic.getName())) {
+            outputRecordsFromTest.put(record.topic.getName(), new ArrayList<>());
+          }
+          outputRecordsFromTest.get(record.topic.getName()).add(
+              new FakeKafkaRecord(record, new ProducerRecord(
+                  record.topic.getName(),
+                  null,
+                  record.timestamp(),
+                  record.key(),
+                  record.topic.getSerializer(schemaRegistryClient).serialize(
+                      record.topic.getName(),
+                      record.value()
+                  )
+              )));
+        }
+    );
+    return outputRecordsFromTest;
+  }
+
+
+  private static FakeKafkaRecord getFakeKafkaRecordFromProducerRecord(
+      final Topic topic,
+      final ProducerRecord producerRecord) {
+    Objects.requireNonNull(producerRecord);
+    final Record testRecord =  new Record(
+        topic,
+        producerRecord.key().toString(),
+        producerRecord.value(),
+        producerRecord.timestamp(),
+        null
+    );
+    return new FakeKafkaRecord(testRecord, producerRecord);
+  }
+
+  private static Topic getTopic(
+      final KafkaTopicClient kafkaTopicClient,
+      final KsqlTopic ksqlTopic,
+      final SchemaRegistryClient schemaRegistryClient) throws IOException, RestClientException {
+    Objects.requireNonNull(kafkaTopicClient);
```

Comment:
instead of requiring nonnull in every one of the private methods, we should just do it once in the public methods